power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp

```diff
@@ -739,49 +757,102 @@ class Deserializer {
         offset_ = msg_data.offset;
         parse_map_array<visit_array_t, move_forward>();
         for (element_number_ = 0; element_number_ != msg_data.size; ++element_number_) {
-            void* element_pointer = component.advance_ptr(scenario_pointer, element_number_);
+            BufferView const element_buffer = advance(buffer, element_number_);
             // check the element is map or array
             auto const element_visitor = parse_map_array<visit_map_array_t, move_forward>();
             if (element_visitor.is_map) {
-                parse_map_element(element_pointer, element_visitor.size, component);
+                parse_map_element(element_buffer, element_visitor.size, component);
             } else {
-                parse_array_element(element_pointer, element_visitor.size, attributes);
+                parse_array_element(element_buffer, element_visitor.size, component, attributes);
             }
         }
         element_number_ = -1;
         offset_ = 0;
     }
 
-    void parse_map_element(void* element_pointer, Idx map_size, MetaComponent const& component) {
+    void parse_map_element(BufferView const& buffer, Idx map_size, MetaComponent const& component) {
         while (map_size-- != 0) {
             attribute_key_ = parse_string();
-            Idx const found_idx = component.find_attribute(attribute_key_);
-            if (found_idx < 0) {
+            Idx const component_attribute_idx = component.find_attribute(attribute_key_);
```

Comment:
The logic here needs to be split, even with a template argument.

For row-based buffer, you try to find the `MetaAttribute` from our internal meta data.

For column-base buffer, you try to find the `MetaAttribute` from the user provided list of columnar buffers.

Reply:
really, there's too much specialization going on here. i'm going to re-implement this entire PR with the knowledge I have now and hope I can write it in such a way that it is more efficient. However, I'm not particularly fond of having to add 8 overloads because for every different combination of `{serializer, deserializer} x {row_based, columnar} x {map, array}` the implementation, caching, buffer mapping and even terminology is completely different and it's quite confusing even for someone like me who has worked with the (de)serializer before. It will be even worse for someone who hasn't.

I'm going to re-implement and I'm first going to spend time on homogenizing the caching mechanism

Reply:
Yes, I agree to re-think about the PR. But still, efficiency/performance is the key here for (de-)serialization, which is more important than the "beauty` of the code. The (de-)serialization is the most complicated and hard to read part anyway (in the sense of C++ features and abstraction). That's almost always the case. We need to accept that fact sometimes...

Reply:
i think i've processed all your comments on the deserializer. if you have time, can you do another check?